File: server/src/breakout/manager.go
package breakout

import "sync"

type Room struct { ID string; Name string; Participants map[string]string }
type Manager struct { mu sync.RWMutex; rooms map[string]*Room }

func NewManager() *Manager { return &Manager{rooms: map[string]*Room{}} }
func (m *Manager) Create(id,name string) *Room { m.mu.Lock(); defer m.mu.Unlock(); r:=&Room{ID:id,Name:name,Participants:map[string]string{}}; m.rooms[id]=r; return r }
func (m *Manager) Assign(roomID,userID string) bool { m.mu.Lock(); defer m.mu.Unlock(); r:=m.rooms[roomID]; if r==nil{return false}; r.Participants[userID]=roomID; return true }
func (m *Manager) Remove(roomID,userID string) bool { m.mu.Lock(); defer m.mu.Unlock(); r:=m.rooms[roomID]; if r==nil{return false}; if _,ok:=r.Participants[userID];!ok{return false}; delete(r.Participants,userID); return true }
func (m *Manager) Get(roomID string) *Room { m.mu.RLock(); defer m.mu.RUnlock(); return m.rooms[roomID] }

File: server/src/breakout/manager.ts
export type BreakoutRoom = {
  id: string;
  parentRoomId: string;
  name: string;
  participants: Record<string, string>;
};

export class BreakoutManager {
  private readonly rooms = new Map<string, BreakoutRoom>();

  create(parentRoomId: string, id: string, name: string): BreakoutRoom {
    const room: BreakoutRoom = {
      id,
      parentRoomId,
      name,
      participants: {}
    };
    this.rooms.set(id, room);
    return this.clone(room);
  }

  assign(parentRoomId: string, breakoutId: string, userId: string): boolean {
    const target = this.rooms.get(breakoutId);
    if (!target || target.parentRoomId !== parentRoomId) return false;
    for (const room of this.rooms.values()) {
      delete room.participants[userId];
    }
    target.participants[userId] = breakoutId;
    return true;
  }

  remove(parentRoomId: string, breakoutId: string, userId: string): boolean {
    const room = this.rooms.get(breakoutId);
    if (!room || room.parentRoomId !== parentRoomId || !(userId in room.participants)) return false;
    delete room.participants[userId];
    return true;
  }

  get(parentRoomId: string, breakoutId: string): BreakoutRoom | undefined {
    const room = this.rooms.get(breakoutId);
    return room && room.parentRoomId === parentRoomId ? this.clone(room) : undefined;
  }

  list(parentRoomId: string): BreakoutRoom[] {
    return [...this.rooms.values()]
      .filter(room => room.parentRoomId === parentRoomId)
      .map(room => this.clone(room));
  }

  private clone(room: BreakoutRoom): BreakoutRoom {
    return { ...room, participants: { ...room.participants } };
  }
}
